### PR TITLE
feat(task): prevent update if task is active

### PR DESCRIPTION
**Pull request by stonebuzz**
prevent update if task is active

internal ref : !27480

### CODE REVIEW

File: inc/taskview.class.php [hunk 1]
```diff
@@ -284,6 +284,20 @@ public function showJobLogs()
...
 
+    /**
+    * Prepare data before update in database
+    *
+    * @param array $input
+    * @return array
+    */
+    public function prepareInputForUpdate($input)
+    {
+        if ($input['is_active'] && $input['is_active_from_db']) {
```
Comment (cedric-anne) [inline inc/taskview.class.php:295]: 1. Should rely on value from DB, it is safer than a values passed via a from.
2. `is_active` may be missing from input, for instance if task is updated via API.

```suggestion
        if ($this->fields['is_active'] && ($input['is_active'] ?? '1')) {
```

File: tests/Integration/Tasks/CronTaskTest.php [hunk 2]
```diff
@@ -638,10 +638,18 @@ public function prepareTaskNotRePrepareIfSuccessful()
...
+        //update task and by deactivating it (plugin prevent task update if it active)
         $pfTask->update([
-         'id'                      => $pfTask->fields['id'],
-         'reprepare_if_successful' => 1,
+          'id'                      => $pfTask->fields['id'],
+          'reprepare_if_successful' => 1,
+          'is_active' => 0,
         ]);
+
+        //update task and by reactivating it
```
Comment (cedric-anne) [inline tests/Integration/Tasks/CronTaskTest.php:648]:
```suggestion
        // reactivate task
```

**Comment by trasher** [comment]
Tests are failing

File: inc/taskview.class.php [hunk 3]
```diff
@@ -284,6 +284,20 @@ public function showJobLogs()
...
 
+    /**
+    * Prepare data before update in database
+    *
+    * @param array $input
+    * @return array
+    */
+    public function prepareInputForUpdate($input)
+    {
+        if (!defined('TU_USER') && $this->fields['is_active'] && ($input['is_active'] ?? '1')) {
```
Comment (cedric-anne) [inline inc/taskview.class.php:295]: Not sure it is a good idea t check for `TU_USER` here.

File: inc/taskview.class.php [hunk 4]
```diff
@@ -284,6 +284,20 @@ public function showJobLogs()
       });");
     }
 
+    /**
+    * Prepare data before update in database
+    *
+    * @param array $input
+    * @return array
+    */
+    public function prepareInputForUpdate($input)
```
Comment (cedric-anne) [inline inc/taskview.class.php:293]: Is this normal that this change is made on `PluginGlpiinventoryTaskView`? I may miss something, but I guess that the data that your are trying to check is related to `PluginGlpiinventoryTask`.

Comment (stonebuzz) [reply inc/taskview.class.php:293]: you're right

File: inc/taskview.class.php (hunk 3 above)
Comment (stonebuzz) [reply inc/taskview.class.php:295]: fixed

File: .phpunit.result.cache
Comment (cedric-anne) [inline .phpunit.result.cache:1]: This file should eb removed.

File: inc/taskview.class.php [hunk 5]
```diff
@@ -284,6 +284,20 @@ public function showJobLogs()
...
+    {
+        if (!defined('TU_USER') && $this->fields['is_active'] && ($input['is_active'] ?? '1')) {
+            Session::addMessageAfterRedirect(__('The task cannot be updated if it is active', 'glpiinventory'), false, ERROR);
+            return false;
+        }
+        return $input;
+    }
```
Comment (cedric-anne) [inline inc/taskview.class.php:300]:
```suggestion
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: inc/taskview.class.php (hunk 5 above)
Comment (stonebuzz) [reply inc/taskview.class.php]: arfff ...
done !

**Review by cedric-anne — APPROVED** [review]

File: tests/Integration/Tasks/CronTaskTest.php [hunk 6]
```diff
@@ -503,11 +503,11 @@ public function prepareTaskNotRePrepareIfSuccessful()
...
         $this->assertArrayHasKey('id', $pfTask->fields);
-        $pfTask->update([
-         'id'                      => $pfTask->fields['id'],
-         'reprepare_if_successful' => 0,
-         'is_active'               => 1
-        ]);
+
+        //update directly in DB to prevent reset jobstate
+        //see PluginGlpiinventoryTask->post_updateItem
+        $DB->query("UPDATE glpi_plugin_glpiinventory_tasks set `reprepare_if_successful` = 0 WHERE id = " . $pfTask->fields['id']);
```
Comment (trasher) [inline tests/Integration/Tasks/CronTaskTest.php:509]: Please avoid using `DB::query()` directly; you can use `DB::update()` here.

File: tests/Integration/Tasks/CronTaskTest.php [hunk 7]
```diff
@@ -638,10 +638,11 @@ public function prepareTaskNotRePrepareIfSuccessful()
         $counters = $data['tasks'][$pfTask->fields['id']]['jobs'][$pfTaskjob->fields['id']]['targets']['PluginGlpiinventoryDeployPackage_' . $pfDeployPackage->fields['id']]['counters'];
         $this->assertEquals($reference, $counters);
 
-        $pfTask->update([
-         'id'                      => $pfTask->fields['id'],
-         'reprepare_if_successful' => 1,
-        ]);
+        //update directly in DB to prevent reset jobstate
+        //see PluginGlpiinventoryTask->post_updateItem
+        $DB->query("UPDATE glpi_plugin_glpiinventory_tasks set `reprepare_if_successful` = 1 WHERE id = " . $pfTask->fields['id']);
```
Comment (trasher) [inline tests/Integration/Tasks/CronTaskTest.php:643]: Use `DB::update()` too

**Review by trasher — CHANGES_REQUESTED** [review]

File: tests/Integration/Tasks/CronTaskTest.php (hunk 6 above)
Comment (stonebuzz) [reply tests/Integration/Tasks/CronTaskTest.php]: done !

File: tests/Integration/Tasks/CronTaskTest.php (hunk 7 above)
Comment (stonebuzz) [reply tests/Integration/Tasks/CronTaskTest.php]: done !

File: .phpunit.result.cache
Comment (trasher) [inline .phpunit.result.cache:1]: Should not have been comited ;)

File: tests/Integration/Tasks/CronTaskTest.php [hunk 8]
```diff
@@ -503,11 +503,15 @@ public function prepareTaskNotRePrepareIfSuccessful()
...
-        $pfTask->update([
-         'id'                      => $pfTask->fields['id'],
-         'reprepare_if_successful' => 0,
-         'is_active'               => 1
-        ]);
+
+        //update directly in DB to prevent reset jobstate
+        //see PluginGlpiinventoryTask->post_updateItem
+        $DB->update(
+            'glpi_plugin_glpiinventory_tasks',
```
Comment (trasher) [inline tests/Integration/Tasks/CronTaskTest.php:510]:
```suggestion
           $pfTask->getTable(),
```

File: tests/Integration/Tasks/CronTaskTest.php [hunk 9]
```diff
@@ -638,10 +642,15 @@ public function prepareTaskNotRePrepareIfSuccessful()
...
         $this->assertEquals($reference, $counters);
 
-        $pfTask->update([
-         'id'                      => $pfTask->fields['id'],
-         'reprepare_if_successful' => 1,
-        ]);
+        //update directly in DB to prevent reset jobstate
+        //see PluginGlpiinventoryTask->post_updateItem
+        $DB->update(
+            'glpi_plugin_glpiinventory_tasks',
```
Comment (trasher) [inline tests/Integration/Tasks/CronTaskTest.php:648]:
```suggestion
            $pfTask->getTable(),
```

**Review by trasher — APPROVED** [review]

**Pull request merged**